### ISSUE TITLE
Is it okay to use teacher forcing in evaluation?

### ISSUE THREAD

**Issue by DaehanKim** [ddddwy/Reinforce-PG#3]
Hi Wu!

I'm studying conditional text generation using your code. Thanks for publishing such a great code.

while reading eval.py, I wondered whether it is allowed to use teacher forcing in evaluation mode, since evaluation process, I think, was supposed to be the same as the model's decoding process. (including beam search)

On the other hand, I suspect it is okay because we use teacher forcing in the training mode. That is, we aim to evaluate its performance in the same environment as the training mode.  

Which one do you think is the convention?  Thanks in advance!

**Comment by wyu-du** [comment]
Hi @DaehanKim,

Thank you for your interest in my work!

The eval.py is used for cross-validation during training, so that we can save the best model and prevent overfitting.

Although you can use beam search in "evaluation mode" (in this case I think "validation mode" may be more suitable), it is time-consuming. So in practice, people usually only apply beam search at the test set. For example, you may refer to [this tutorial](https://github.com/bentrevett/pytorch-seq2seq/blob/master/4%20-%20Packed%20Padded%20Sequences%2C%20Masking%2C%20Inference%20and%20BLEU.ipynb).

Best,
Wanyu

**Comment by DaehanKim** [comment]
Wanyu, Thank you for the reply!

How about using teacher forcing in evaluation mode?
I saw the tutorial you mentioned, but they also turned off teacher forcing when they evaluate.
I wonder whether this part was intended :

![image](https://user-images.githubusercontent.com/20675681/81133224-f1e89880-8f8b-11ea-9264-285298bbccf5.png)

Best regards,
Daehan

**Comment by wyu-du** [comment]
Hi Daehan,

The tutorial didn't turn off teacher forcing in the evaluation mode (their comment is misleading), as you can see, they still feed ground truth words as the input to the model in order to compute the loss.

And the screenshot here is just doing the same thing as the tutorial.

Best,
Wanyu

**Comment by DaehanKim** [comment]
Thank you for your help!
